### PR TITLE
Fix rendering of cached simple marker images on hidpi map displays

### PR DESCRIPTION
**Pull request by nyalldawson**

### CODE REVIEW

File: src/core/symbology/qgsmarkersymbollayer.cpp [hunk 1]
```diff
@@ -1178,7 +1178,11 @@ bool QgsSimpleMarkerSymbolLayer::prepareCache( QgsSymbolRenderContext &context )
     return false;
   }
 
-  mCache = QImage( QSize( imageSize, imageSize ), QImage::Format_ARGB32_Premultiplied );
+  mCache = QImage( QSize( imageSize * context.renderContext().devicePixelRatio(),
```
Comment (elpaso) [inline src/core/symbology/qgsmarkersymbollayer.cpp:1181]: Shouldn't this new imageSize be tested for MAXIMUM_CACHE_WIDTH ? I.e. move the check after the image size is multiplied by devicePixelRatio ?

Comment (nyalldawson) [reply src/core/symbology/qgsmarkersymbollayer.cpp]: Correct, fixed

**Review by nirvn — APPROVED** [review]

File: src/core/symbology/qgsmarkersymbollayer.cpp [hunk 2]
```diff
@@ -2325,7 +2330,8 @@ void QgsSvgMarkerSymbolLayer::renderPoint( QPointF point, QgsSymbolRenderContext
 
   bool hasDataDefinedSize = false;
   const double scaledWidth = calculateSize( context, hasDataDefinedSize );
-  const double width = context.renderContext().convertToPainterUnits( scaledWidth, mSizeUnit, mSizeMapUnitScale );
+  const double devicePixelRatio = context.renderContext().devicePixelRatio();
+  const double width = context.renderContext().convertToPainterUnits( scaledWidth, mSizeUnit, mSizeMapUnitScale ) * devicePixelRatio;
```
Comment (nirvn) [inline src/core/symbology/qgsmarkersymbollayer.cpp:2334]: @nyalldawson , I've tested the PR, and this line is problematic as it doubles the size of SVG markers when rendering from a non-cached QPicture. The width should only be doubled when rendering from the cached QImage.

Comment (nyalldawson) [reply src/core/symbology/qgsmarkersymbollayer.cpp]: Should be fixed now

**Pull request merged**